enhancement: generalising docker build and run for multicloud environ…

raflFaisal opened the pull request:
This PR introduces a flexible and extensible deployment interface for the adk deploy command to support multiple cloud providers (e.g., GCP, AWS, Kubernetes, etc.) and Docker environments.

This proposed changes improves the adk deploy interface to be more flexible, and scalable, enabling easier integration with various cloud environments and future cloud provider support. For more details please refer this https://github.com/google/adk-python/issues/184

Key Changes:
Generalised Dockerfile:
Removed provider-specific hard-coded variables.

Introduced two options for injecting environment variables:
(1) Using .env files for local development (docker).
(2)Passing environment variables directly via the adk deploy command for production environments.

Modular Deployment Interface:
(1) Introduced --cloud-provider to specify the target platform (default: gcp).
(2) Introduced --provider-args for provider-specific parameters (e.g., aws-region, project).
(3) Implemented cloud-specific deployers for each platform (e.g., GCPDeployer, DockerDeployer).

Used a factory pattern to route deployment logic based on the cloud provider.
Simplified deployment commands with clear, flexible syntax for docker, GCP (AWS, k8s and other cloud providers in future)
Readme.md file is updated with usage instructions

What Will Not Be Supported:
(1) Non-containerized deployments (e.g., serverless functions).
(2) CI/CD integration.
(3) Credential management (assumes provider CLIs are pre-configured).

Test Plan:
Unit Tests:
✅ Existing unittest cases adopted to test deployment logic for 'cloud run'
✅ implemented unitest case for Docker environment handling.

Manual Verification: Successfully deployed and verified agent behavior using the following commands:
✅ adk deploy cloud_run agents/llm_auditor
✅ adk deploy cloud_run --with_ui agents/llm_auditor
✅ adk deploy cloud_run --project=myproject-9094 --region=europe-west3 agents/llm_auditor
✅ adk deploy cloud_run --project=myproject-9094 --region=europe-west3 --with_ui agents/llm_auditor
✅ adk deploy cloud_run --project=myproject-9094 --region=europe-west3 --with_ui
--env GOOGLE_GENAI_USE_VERTEXAI=0
--env GOOGLE_API_KEY=secret-key
agents/llm_auditor

✅ adk deploy docker agents/llm_auditor
✅ adk deploy docker --with_ui agents/llm_auditor
✅ adk deploy docker --with_ui
--env GOOGLE_GENAI_USE_VERTEXAI=0
--env GOOGLE_API_KEY=secret-key
agents/llm_auditor


Note: This PR is a refreshed version of [#395](https://github.com/google/adk-python/pull/395), which was closed due to staleness during repository cleanup. All feedback from the previous thread has been addressed.

raflFaisal committed:
enhancement: generalising docker build and run for multicloud environments

raflFaisal committed:
incorporating review comments by gemini

raflFaisal committed:
remove comma left as typo

raflFaisal committed:
fix pyink reported issues

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

glyphiglyphi commented:
Wow this is what I have been looking for! 
please @seanzhou1023 @hangfei @DeanChensj move this forward 🙏

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

DeanChensj commented on src/google/adk/cli/cli_tools_click.py:
```diff
@@ -1426,12 +1456,84 @@ def cli_deploy_cloud_run(
+@deploy.command("docker", cls=HelpfulCommand)
+@add_common_deploy_options
+@adk_services_options()
+@deprecated_adk_services_options()
```
we can remove the @deprecated_adk_services_options here

raflFaisal replied:
Thanks for the nice suggestion, I have removed the deprecated_adk_services_options !

DeanChensj commented on src/google/adk/cli/deployers/base_deployer.py:
Does everything in the deployers folder need to depend on a docker template / image?  Are we able to also generalize agent_engine as one of the deployer?  

Do we need to add some README or comment regarding what is needed to add a new deployer option?

raflFaisal replied:
The base Deployer is platform-agnostic. Current implementations (DockerDeployer, CloudRunDeployer) use Docker as a common packaging format. I noticed that GKE and agent_engine deployments were added after this PR, GKE also uses the same Docker template, which I can generalize in a separate MR (with proper testing). I’ll also explore whether agent_engine can be integrated into the deployers factory in a future PR.

DeanChensj replied:
Thanks, that sounds like a perfect plan.  Could you add the README in this PR, and we can add other improvement in future PRs.

raflFaisal replied:
Sure! I have added a README under deployers folder with some notes and instructions. Let me know if it looks good to you. Thanks !

raflFaisal replied:
Hi @DeanChensj @glyphiglyphi,
Just a gentle follow-up, merging this PR would unblock my work on generalising GKE deployment :)
Appreciate your time and review. Thanks!

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

faisalBooking committed:
review comments: incorporating suggestions

raflFaisal commented on src/google/adk/cli/cli_tools_click.py:
```diff
@@ -115,6 +116,124 @@ def main():
   pass
 
 
+def add_common_deploy_options(command):
```
Sure, deploy_options makes more sense !

raflFaisal commented:
I’m updating the test cases to align with the recent changes. I’ll provide an update once the new test cases have been adopted. Thanks for your support!

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

hangfei committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

faisalBooking committed:
feat: incorporating review comments, adding testcases for deployers

raflFaisal commented on src/google/adk/cli/deployers/cloud_run_deployer.py:
Good idea! I added tests for the deployers class and their helper functions

DeanChensj committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Add copyright and license information to base_deployer.py

Added copyright notice and licensing information.

raflFaisal committed:
Add copyright and license information

Added copyright notice and license information to the file.

raflFaisal committed:
Add copyright and license information

Added copyright notice and licensing information to the file.

raflFaisal committed:
Add copyright and license to docker_deployer.py

Added copyright notice and license information to docker_deployer.py

raflFaisal committed:
Add copyright and license information to dockerfile_template.py

Added copyright notice and license information to the dockerfile_template.py file.

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal commented:
@DeanChensj @hangfei Thanks for the approval! The test cases have been aligned with the new changes.
Please let me know if there’s anything else needed from my side before merging this PR into main.

raflFaisal commented:
Thanks for the update, @DeanChensj. I’ll wait for the internal review and am happy to help if anything’s needed.

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Fix code conflicts

Fix code conflicts due to new code changes pushed

raflFaisal committed:
Resolve code conflicts

Resolve code conflicts

raflFaisal committed:
Resolve code conflicts

Resolve code conflicts

raflFaisal commented:
Hey @DeanChensj, just checking in to see if there are any updates from the internal review.
Happy to make any changes if needed. Thanks!

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal committed:
Merge branch 'main' into raflFaisal/generaliseContainerEnvForAdkAgent

raflFaisal commented:
Hi @DeanChensj and @hangfei,

Hope you're doing well! I wanted to check in on this PR and see whether there have been any updates from the internal review process.

I've been keeping an eye on changes in ADK and am happy to rebase, resolve conflicts, or make any other adjustments that would help move this forward. I'm also open to adapting the implementation if there are any concerns around scope or design that have come up during review.

Thank you again for all the time and feedback that has already gone into reviewing this contribution. I appreciate the collaboration and would be glad to help in any way I can.
